File: library/src/main/java/com/github/alexjlockwood/kyrie/GroupNode.java
package com.github.alexjlockwood.kyrie;

import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.PointF;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

/** A {@link Node} that holds a group of children {@link Node}s. */
public final class GroupNode extends BaseNode {
  @NonNull private final List<Node> children;

  private GroupNode(
      List<Animation<?, Float>> rotation,
      List<Animation<?, Float>> pivotX,
      List<Animation<?, Float>> pivotY,
      List<Animation<?, Float>> scaleX,
      List<Animation<?, Float>> scaleY,
      List<Animation<?, Float>> translateX,
      List<Animation<?, Float>> translateY,
      List<Node> children) {
    super(rotation, pivotX, pivotY, scaleX, scaleY, translateX, translateY);
    this.children = children;
  }

  @NonNull
  List<Node> getChildren() {
    return children;
  }

  // <editor-fold desc="Layer">

  @NonNull
  @Override
  GroupLayer toLayer(PropertyTimeline timeline) {
    return new GroupLayer(timeline, this);
  }

  private static class GroupLayer extends BaseLayer {
    @NonNull private final List<Layer> children;

    public GroupLayer(PropertyTimeline timeline, GroupNode node) {
      super(timeline, node);
      final List<Node> childrenNodes = node.getChildren();
      children = new ArrayList<>(childrenNodes.size());
      for (int i = 0, size = childrenNodes.size(); i < size; i++) {
        children.add(childrenNodes.get(i).toLayer(timeline));
      }
    }

    @Override
    public void onDraw(Canvas canvas, Matrix parentMatrix, PointF viewportScale) {
      canvas.save();
      for (int i = 0, size = children.size(); i < size; i++) {
        children.get(i).draw(canvas, parentMatrix, viewportScale);
      }
      canvas.restore();
    }
  }

  // </editor-fold>

  // <editor-fold desc="Builder">

  public static Builder builder() {
    return new Builder();
  }

  /** Builder class used to create {@link GroupNode}s. */
  public static final class Builder extends BaseNode.Builder<Builder> {
    private final List<Node> children = new ArrayList<>();

    private Builder() {}

    // Children.

    public Builder child(Node node) {
      children.add(node);
      return this;
    }

    public Builder child(Node.Builder builder) {
      return child(builder.build());
    }

    @NonNull
    @Override
    Builder self() {
      return this;
    }

    @NonNull
    public GroupNode build() {
      return new GroupNode(
          rotation, pivotX, pivotY, scaleX, scaleY, translateX, translateY, children);
    }
  }

  // </editor-fold>
}

File: kyrie/src/main/java/com/github/alexjlockwood/kyrie/GroupNode.kt
package com.github.alexjlockwood.kyrie

import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.PointF

/** A [Node] that holds a group of children [Node]s. */
class GroupNode private constructor(
        rotation: List<Animation<*, Float>>,
        pivotX: List<Animation<*, Float>>,
        pivotY: List<Animation<*, Float>>,
        scaleX: List<Animation<*, Float>>,
        scaleY: List<Animation<*, Float>>,
        translateX: List<Animation<*, Float>>,
        translateY: List<Animation<*, Float>>,
        private val children: List<Node>
) : TransformNode(rotation, pivotX, pivotY, scaleX, scaleY, translateX, translateY) {

    // <editor-fold desc="Layer">

    override fun toLayer(timeline: PropertyTimeline): GroupLayer {
        return GroupLayer(timeline, this)
    }

    internal class GroupLayer(timeline: PropertyTimeline, node: GroupNode) : TransformNode.TransformLayer(timeline, node) {
        private val children: ArrayList<Layer>

        init {
            val childrenNodes = node.children
            children = ArrayList(childrenNodes.size)
            var i = 0
            val size = childrenNodes.size
            while (i < size) {
                children.add(childrenNodes[i].toLayer(timeline))
                i++
            }
        }

        override fun onDraw(canvas: Canvas, parentMatrix: Matrix, viewportScale: PointF) {
            canvas.save()
            children.forEach { it.draw(canvas, parentMatrix, viewportScale) }
            canvas.restore()
        }

        override fun isStateful(): Boolean {
            for (i in 0 until children.size) {
                if (children[i].isStateful()) {
                    return true
                }
            }
            return false
        }

        override fun onStateChange(stateSet: IntArray): Boolean {
            var changed = false
            for (i in 0 until children.size) {
                changed = changed or children[i].onStateChange(stateSet)
            }
            return changed
        }
    }

    // </editor-fold>

    // <editor-fold desc="Builder">

    @DslMarker
    private annotation class GroupNodeMarker

    /** Builder class used to create [GroupNode]s. */
    @GroupNodeMarker
    class Builder internal constructor() : TransformNode.Builder<Builder>() {
        private val children = ArrayList<Node>()

        // Children.

        fun child(node: Node): Builder {
            children.add(node)
            return this
        }

        fun child(builder: Node.Builder<*>): Builder {
            return child(builder.build())
        }

        override val self = this

        override fun build(): GroupNode {
            return GroupNode(rotation, pivotX, pivotY, scaleX, scaleY, translateX, translateY, children)
        }
    }

    // </editor-fold>

    companion object {

        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }
    }
}
